File: Q011/Q011.py
import os
import numpy as np

def product_four(lst):
    return max(reduce(lambda x,y:x*y, lst[i:i+4]) for i in range(len(lst)-3))

def hproduct(array):
    result = 0
    for row in array:
        result = max(result, product_four(row))
    return result

def dproduct(array):
    """ 
    this function only find the four consecutive number from upper left to lower right.
    If the other diagonal direction is the target, just rotate the array will be fine.
    """
    result = 0
    for i in range(4, len(array)):
        result = max(result,  product_four(array[:i, :i].diagonal())\
                     + product_four(array[-i:, :i].diagonal()))
    return result

def largest_product(array):
    return max([hproduct(array), hproduct(array.T), dproduct(array), dproduct(np.rot90(array))])

def main():
    # the size of array is pre-defined in the question
    array = np.zeros((20,20), dtype=int)
    ln = 0
    with open(os.path.join(os.path.dirname(__file__), "Q011.txt")) as datafile:
        for row in datafile:
            array[ln] = [int(k) for k in row.split(' ')]
            ln += 1
    print largest_product(array)

if __name__ == "__main__":
    main()

File: Q011/Q011.rb
load "../matrix_extend.rb"

# the produce could be splitted into three parts:
# 1. horizontal and vertical -- horizontal products for matrix and its rotation
# 2. lower-left for both matrix and its rotation
# 3. upper-right for both matrix and its rotation
pro_length = 4
m = Matrix.from_a(File.read("Q011.txt").split("\n").map{|i| i.split(" ").map{|k| k.to_i}})
m90 = m.rotate
side = m.row_size
result = 0

side.times do |i|
  acc = (0..(side-4)).map{|j| m.row(i)[j..(j+3)].reduce(:*)}
  acc += (0..(side-4)).map{|j| m90.row(i)[j..(j+3)].reduce(:*)}

  less_row = m.minor(i..side,0..side)
  less_col = m.minor(0..side,i..side)
  less_row90 = m90.minor(i..side,0..side)
  less_col90 = m90.minor(0..side,i..side)
  if less_row.row_size >= 4
    acc += (0..(less_row.row_size-4)).map{|j| less_row.diagonal[j..(j+3)].reduce(:*)}
    acc += (0..(less_row.row_size-4)).map{|j| less_row90.diagonal[j..(j+3)].reduce(:*)}
  end
  if less_col.row_size >= 4
    acc += (0..(less_col.column_size-4)).map{|j| less_col.diagonal[j..(j+3)].reduce(:*)}
    acc += (0..(less_col.column_size-4)).map{|j| less_col90.diagonal[j..(j+3)].reduce(:*)}
  end
  result = [acc.max, result].max
end

puts result


